### PR TITLE
Baseline for Keycloak deployment in operator

### LINKED ISSUE #9171: Keycloak.X deployment
The aim of this task is to provide a baseline for deploying Keycloak.X using the operator. This includes:
- [ ] Keycloak CRD
- [ ] Basic scaling options (number of replicas)
- [ ] Basic health and readiness checks

### LINKED ISSUE #9552: Readiness and Liveness probe for the operator deployment of Keycloak.X
### Description

The new Keycloak.X expose some basic health endpoints, we should investigate and model on top of those proper default readiness and liveness probes for the Keycloak deployment.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by vmuzikar**
Limitations that will be addressed as a follow-up:
* No clustering support (#9730)
* Pod Templates are not implemented in the controller (#9729)
* No tests (#9174)

Resolves #9171, resolves #9552

### CODE REVIEW

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 1]
```diff
@@ -16,82 +16,162 @@
...
+            distConfig.putAll(keycloakCR.getSpec().getDistConfig());
+        }
 
-    KeycloakDeployment(KubernetesClient client) {
-        this.client = client;
+        Set<String> configSecretsNames = new HashSet<>();
+        List<EnvVar> configEnvVars = distConfig.entrySet().stream()
+                .map(e -> {
+                    EnvVarBuilder builder = new EnvVarBuilder().withName(e.getKey());
+                    Matcher matcher = CONFIG_SECRET_PATTERN.matcher(e.getValue());
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:92]: ok, I think I got how you want this to work @vmuzikar , can you elaborate on the reasoning for spinning out a custom "vault" integration on top of Kubernetes when the same functionality is offered by `envFrom`?

File: operator/src/main/resources/example-keycloak.yml [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
+apiVersion: keycloak.org/v2alpha1
 kind: Keycloak
 metadata:
   name: example-kc
 spec:
   instances: 1
+  distConfig:
+    KC_DB: postgres
+    KC_DB_URL_HOST: postgres-db
+    KC_DB_USERNAME: ${secret:keycloak-db-secret:username}
```
Comment (andreaTP) [inline operator/src/main/resources/example-keycloak.yml:10]: this functionality is already offered by Kubernetes `envFrom`, I do believe that users are used to the Kubernetes API already and we should provide functionalities with the minimum possible attrition and possibly avoiding new and custom concepts.

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Stores a set of resource names that are related to (but not owned by) some CR. To be used by a controller for watching
+ * those resources. Also stores a {@code modified} flag for each resource. Motivation for this in-memory store is to avoid
+ * touching (e.g. adding labels) the resources that are not owned by the operator and not to store state in K8s.
```
Comment (andreaTP) [inline operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:31]: Adding labels is completely fine IMHO.

**Review by andreaTP — CHANGES_REQUESTED** [review]
First round of review finished!

Great work @vmuzikar !

I like the overall structure and especially the State management, although I think we should try to avoid introducing (so early!) complexities like `WatchedResourcesStore` and the custom "Vault" integration relying respectively on the capabilities of the Java SDK and the standard K8s API.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Represents a single K8s resource that is managed by this operator (e.g. Deployment, Service, Ingress, etc.)
+ *
+ * @author Vaclav Muzikar <vmuzikar@redhat.com>
+ */
+public abstract class OperatorManagedResource {
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:37]: It is overlapping for sure, but since "Dependent Resources" is now still just a [PR](https://github.com/java-operator-sdk/java-operator-sdk/pull/785) and not something we can use right now, I decided to use an interim solution. We can (and we should!) refactor later.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    protected abstract HasMetadata getReconciledResource();
+
+    public void createOrUpdateReconciled() {
+        HasMetadata resource = getReconciledResource();
+        setDefaultLabels(resource);
+        setOwnerReferences(resource);
+
+        Log.debugf("Creating or updating resource: %s", resource);
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:53]: It is still the same logging framework coming from Quarkus. It's just using static methods and not CDI. I realized CDI approach is pretty stupid for logging as we don't have everything as beans and passing deps manually is a mess. This is an elegant way offered by Quarkus and we should stick with it everywhere.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    protected void setDefaultLabels(HasMetadata resource) {
+        Map<String, String> labels = Optional.ofNullable(resource.getMetadata().getLabels()).orElse(new HashMap<>());
+        labels.putAll(Constants.DEFAULT_LABELS);
+        resource.getMetadata().setLabels(labels);
+    }
+
+    protected void setOwnerReferences(HasMetadata resource) {
+        if (!cr.getMetadata().getNamespace().equals(resource.getMetadata().getNamespace())) {
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: > Cross-namespace owner references are disallowed by design. Namespaced dependents can specify cluster-scoped or namespaced owners.

https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+ *
+ * @author Vaclav Muzikar <vmuzikar@redhat.com>
+ */
+public class WatchedResourcesStore {
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:35]: Do we? Could you please point me to it? :) I'd be more than happy to remove this class and rely only on what the SDK provides but AFAIK we currently need to handle ourselves what resources do we watch and how.

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    // namespace --> CR name --> resource name --> modified?
+    Map<String, Map<String, Map<String, Boolean>>> resources = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:37]: I'm open to suggestions. ;) But in general, I don't think we need e.g. generics in this case. For watching resources we basically need just the name as each "store" instance should be used for a single resource type in any case.

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        return Collections.unmodifiableSet(getResources(crNamespace, crName).keySet());
+    }
+
+    /**
+     * Removes all related resources for given CR.
+     *
+     * @param crNamespace
+     * @param crName
+     */
+    public void removeResourcesForCR(String crNamespace, String crName) {
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:126]: The `WatchedResourcesStore` is meant for resources that we don't own. E.g. the Secrets with credentials. So I'd argue that we should use owner references with such objects.

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    /**
+     * Set the modified flag to {@code false} for all resources that are related to given CR.
+     * 
+     * @param crNamespace
+     * @param crName
+     * @return true if there were some modified resources, otherwise false
+     */
+    public boolean resetModified(String crNamespace, String crName) {
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:58]: This method is used by the reconciliation loop to basically check if any of the associated resources have been modified. Additionally, it resets the modified flag on those resources. I might come up with a better name or split it into two methods. ;)

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java [hunk 11]
```diff
@@ -18,59 +18,139 @@
...
+        return Optional.of(kc);
+    }
+
+    @Override
+    public DeleteControl cleanup(Keycloak kc, Context context) {
+        String kcName = kc.getMetadata().getName();
+        String namespace = kc.getMetadata().getNamespace();
+        Log.infof("--- Cleaning up Keycloak: %s in namespace: %s", kcName, namespace);
+
+        watchedSecrets.removeResourcesForCR(namespace, kcName);
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:152]: We are not deleting any resources. Simply removing them from the watched store so that the operator won't trigger reconciliation loop on them.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 12]
```diff
@@ -16,82 +16,162 @@
...
+import java.util.stream.Collectors;
+
+public class KeycloakDeployment extends OperatorManagedResource {
+
+    public static final Pattern CONFIG_SECRET_PATTERN = Pattern.compile("^\\$\\{secret:([^:]+):(.+)}$");
+
+    private final Config config;
+    private final Keycloak keycloakCR;
+    private final Deployment existingDeployment;
+    private final Deployment defaultDeployment;
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:52]: Nope. ;) It is basically the template, the new deployment if it does not exist already. Desired deployment is basically a merge of existing and default. I'm open to suggestions for better naming.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:92]: The motivation behind this is that I did not want to tie the dist config to env vars, at least not from the API PoV that we provide in the CRD. What if we decide in the future that we want to implement it differently and use e.g. config files and not env vars? I was even wondering if we should use some "translation" from Quarkus to allow users to use the config keys as in config files or system properties (the dot notation) as I see it more generic and not the env vars notation (capitals and snake case).

File: operator/src/main/resources/base-keycloak-deployment.yaml [hunk 13]
```diff
@@ -18,23 +14,29 @@ spec:
...
         name: keycloak
         ports:
         - containerPort: 8443
           protocol: TCP
         - containerPort: 8080
           protocol: TCP
+        livenessProbe:
+          exec:
+            command:
+              - curl --head --fail --silent http://127.0.0.1:8080/health/live
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/resources/base-keycloak-deployment.yaml:32]: Already linked. ;)

File: operator/src/main/resources/example-keycloak.yml (hunk 2 above)
Comment (vmuzikar) [reply operator/src/main/resources/example-keycloak.yml:10]: See the comment above please. I wanted to "hide" the impl details from users – that we're using env vars currently

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java [hunk 14]
```diff
@@ -18,59 +18,139 @@
...
+                deploymentInformer, d -> {
+                    List<OwnerReference> ownerReferences = d.getMetadata().getOwnerReferences();
+                    if (!ownerReferences.isEmpty()) {
+                        return Set.of(new ResourceID(ownerReferences.get(0).getName(), d.getMetadata().getNamespace()));
+                    } else {
+                        return Collections.emptySet();
+                    }
+        });
+
+        // TODO optimize this (set labels?) to not watch all namespaces
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:83]: We could use labels of course. But it'd add additional complexity (to already complex problem ;)) – we need to remove those labels once no longer used in any CR etc. Wondering, if the "Dependent Resources" in the SDK could help with this.

**Comment by vmuzikar** [comment]
@andreaTP Thank you for your review! I hopefully answered all the comments. :)

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:37]: deal 👍

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:53]: Do we mind that, in this same codebase, the Keycloak Quarkus distribution takes a different approach?

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:35]: cc. @csviri

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:37]: I'm arguing that we possibly don't need this in-memory representation at all.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: In this case I would `throw` an exception instead of silently ignoring.

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:126]: > E.g. the Secrets with credentials

if we don't own them we should not, as well, remove those.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:83]: > we need to remove those labels once no longer used in any CR etc.

why?
I think we can rely on one or more ["Recommended Labels"](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/)

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:58]: I do like the split into two 👍

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:152]: The approach with `labels` should fix this as well, I think 🙂

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:52]: Naming is hard ... 🙂 
I give it a try:

`existing` -> `current`
`default` -> `base`

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:92]: I was thinking about this indirection and I do believe that it brings a number of downsides:

- not standard
- it works as a re-name that can be misleading
- the re-naming makes it harder to debug issues
- it's not operator business to mangle Keycloak distribution properties
- custom implementation of an already existing feature
- no proper feedback for the user in case of misconfiguration
- subject to subtle and nasty bugs (e.g. add a space `${ source:`)
- additional logic to be handled

I do see a single interesting use-case that, potentially(not with the current implementation), this feature can bring on the table that is:

 - fetch env variable values from external Secrets (such as the credentials of the DB from the Secret created by the Crunchy operator)
 
 to get there we would need to add support for namespaces and Json parsing.
 
All in all, this PR is intended to be a `Baseline` for the deployment, I believe we agree that this is already a `feature`, so, my suggestion is to leave the passwords in clear text for now, remove this logic, and open a issue to re-visit/re-iterate/better discuss this feature in a separate PR.

File: operator/src/main/resources/example-keycloak.yml (hunk 2 above)
Comment (andreaTP) [reply operator/src/main/resources/example-keycloak.yml:10]: commented above as well 👍

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:83]: @vmuzikar heads up since I didn't knew about, but a possible approach that we can explore is to have a Reconciler for those `Secrets`, as described in the section `Support for Well Known (non-custom) Kubernetes Resources` in https://javaoperatorsdk.io/docs/features

File: operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (csviri) [reply operator/src/main/java/org/keycloak/operator/WatchedResourcesStore.java:35]: Hi, yes informers are caching the resources, you generally have everything if not managing non K8S resources:

https://github.com/java-operator-sdk/java-operator-sdk/blob/e5c3c713d7431c246f5f4775594dbd45324447eb/sample-operators/tomcat-operator/src/main/java/io/javaoperatorsdk/operator/sample/TomcatReconciler.java#L61

see this, but if you just create a reference in the reconciler to the informer event source, you can access it's api directly, it has bunch of methods to access cache and cached objects

**Comment by vmuzikar** [comment]
I removed the watched resources store (i.e. the Deployment won't be restarted on secret change). But I decided not to change the format for referring Secrets (`${secret:...`) for now as it would create even bigger difference between local configuration using conf files and configuration using CRs. Because we'd need to use name and values separately.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java [hunk 15]
```diff
@@ -18,59 +18,101 @@
...
+
+        Log.infof("--- Reconciling Keycloak: %s in namespace: %s", kcName, namespace);
 
+        var statusBuilder = new KeycloakStatusBuilder();
+
+        // TODO use caches in secondary resources; this is a workaround for https://github.com/java-operator-sdk/java-operator-sdk/issues/830
+        // KeycloakDeployment deployment = new KeycloakDeployment(client, config, kc, context.getSecondaryResource(Deployment.class).orElse(null));
+        var kcDeployment = new KeycloakDeployment(client, config, kc, null);
+        kcDeployment.updateStatus(statusBuilder);
+        kcDeployment.createOrUpdateReconciled();
```
Comment (andreaTP) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: I might be missing something, are we recreating the deployment all of the times?

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 16]
```diff
@@ -16,82 +16,162 @@
...
-            baseDeployment = client.apps().deployments().load(url).get();
+    public void updateStatus(KeycloakStatusBuilder status) {
+        if (existingDeployment == null) {
+            status.addNotReadyMessage("No existing Deployment found, waiting for creating a new one");
+            return;
         }
 
-        var deployment = baseDeployment;
+        var replicaFailure = existingDeployment.getStatus().getConditions().stream()
+                .filter(d -> d.getType().equals("ReplicaFailure")).findFirst();
```
Comment (andreaTP) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:145]: Just out of curiosity, any reason why are you re-mapping just this specific failure condition?

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 17]
```diff
@@ -16,82 +16,162 @@
...
+            status.addErrorMessage("Deployment failure: " + replicaFailure.get());
+            return;
+        }
 
-        deployment
-                .getSpec()
-                .setReplicas(keycloak.getSpec().getInstances());
+        if (existingDeployment.getStatus() == null
+                || existingDeployment.getStatus().getReadyReplicas() == null
+                || existingDeployment.getStatus().getReadyReplicas() < keycloakCR.getSpec().getInstances()) {
```
Comment (andreaTP) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:154]: We might want to improve a little on this in the future with something on those lines:

```java
existingDeployment.getStatus().getReadyReplicas() == existingDeployment.getSpec().getReplicas() &&
existingDeployment.getStatus().getReadyReplicas() == keycloakCR.getSpec().getInstances() &&
existingDeployment.getStatus().getUnavailableReplicas() == 0
```
but is good enough for now.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java [hunk 18]
```diff
@@ -16,9 +16,16 @@
...
 
+import io.fabric8.kubernetes.api.model.PodTemplate;
+
+import java.util.Map;
+
 public class KeycloakSpec {
 
     private int instances = 1;
+    private String image;
+    private Map<String, String> distConfig;
```
Comment (andreaTP) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java:27]: not sure about the name, for Keycloak users is always a `distribution`, maybe just `configuration`?

File: operator/src/main/resources/postgres.yaml [hunk 19]
```diff
@@ -0,0 +1,39 @@
+# PostgreSQL StatefulSet
```
Comment (andreaTP) [inline operator/src/main/resources/postgres.yaml:1]: should we name this file `example-postgres` or `demo-postgres` ?

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:37]: can you mark this as resolved @vmuzikar

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:53]: this is still valid, but I don't have a super strong opinion against, feel free to mark this as `Resolved` if we don't care about consistency in this codebase.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: this comment instead is still valid, can you, please, either remove the condition at all or throw an Exception instead of silently returning?

**Review by andreaTP — CHANGES_REQUESTED** [review]
@vmuzikar Great update! 👏 

A few more comments, but only one important regarding the(missing or I missed it 🙂 ) condition for recreating the Deployment, all the rest is totally in the right direction!

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 15 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: Nope, resource version is not changed most of the times this is called. As far as I understand it, it's something like `kubectl apply`. Not sure if we like [`edit`](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#deployment) more. ;)

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 16 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:145]: Unfortunately, conditions are missing official documentation but to my understanding it is a good practice that K8s API server creates this condition when there are errors. Other conditions just tell us if we have all pods ready or if they're scaling. This one is about errors specifically.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 17 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:154]: Do we care about unavailable replicas and all replicas? I think the only thing that concerns us is the number of ready replicas.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java (hunk 18 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java:27]: `configuration` is too generic and could imply also the deployment configuration. `distribution` is also too generic, what would the field do? How about `distributionConf`? Or `distributionConfiguration`?

File: operator/src/main/resources/postgres.yaml (hunk 19 above)
Comment (vmuzikar) [reply operator/src/main/resources/postgres.yaml:1]: 👍

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:53]: Well, AFAIK rest of the KC can't fully utilize Quarkus features (due to WF), that includes the logging. In the operator we can use what Quarkus provides.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: What if we have a managed resource in different namespace than the CR? We surely want to support those scenarios. Of course, the resources then would have to be "manually" removed by the Operator upon CR deletion.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 15 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: I do not like `edit` more ...

But according to this [comment](https://github.com/fabric8io/kubernetes-client/issues/2454#issuecomment-708316924) the resource should always be updated.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 16 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:145]: Thanks for explaining!

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 15 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: I'm not saying there's no API request with each call but if the resource is not effectively updated, its version remains the same. That's fine from my PoV, at least for now.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 17 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:154]: not fussy on this, totally good enough especially for the baseline 👍

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java (hunk 18 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java:27]: Not super sure, but I would pick `distributionConfiguration` in absence of anything better.

Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java:27]: Maybe `serverConf`?

Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java:27]: For a public API I would prefer to use avoid abbreviations(up to a point 🙂  )

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 15 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: I was looking for a method like `boolean shouldReconcile()` in `KeycloakDeployment`, but we can always add it later.
If you tested this a bit, I'm fine 👍

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java (hunk 18 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java]: Ok, renamed to `serverConfiguration`. 😉

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: correct, but the early return doesn't solve this either ...

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java (hunk 18 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java]: ok 👍

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 15 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: Sure, let's keep it simple for the prototype for now and watch for possible issues. We can always add a complex logic for comparing and evaluating the need of calling `createOrUpdate`.

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: So what do you suggest exactly? For sure throwing an exception is not a solution in context of this method. Of course, for true support for foreign resources located in other namespaces we'd need to implement additional logic.

In any case, I'm pretty sure this gets refactored sooner or later when the Dependent Resources are ready in the SDK.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java (hunk 15 above)
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:92]: Deal 👍

File: operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/OperatorManagedResource.java:65]: I would completely remove this condition as it's misleading and not used ATM, but not anything blocking 👍

**Review by andreaTP — DISMISSED** [review]
LGTM, totally good enough to start iterating! 💯

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 20]
```diff
@@ -16,82 +16,162 @@
...
+            Log.info("Trying to fetch existing Deployment from the API");
+            this.existingDeployment = fetchExistingDeployment();
+        }
+
+        baseDeployment = createBaseDeployment();
     }
 
-    private Deployment baseDeployment;
+    @Override
+    protected HasMetadata getReconciledResource() {
```
Comment (jonathanvila) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:64]: Probably using other variables names could make this less confusing ? wdyt ?

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java [hunk 21]
```diff
@@ -16,9 +16,16 @@
...
+import io.fabric8.kubernetes.api.model.PodTemplate;
+
+import java.util.Map;
+
 public class KeycloakSpec {
 
     private int instances = 1;
+    private String image;
+    private Map<String, String> serverConfiguration;
+    private PodTemplate podTemplate;
```
Comment (jonathanvila) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java:28]: I would suggest to move this out to another PR

File: operator/src/main/resources/application.properties [hunk 22]
```diff
@@ -2,3 +2,6 @@ quarkus.operator-sdk.crd.apply=true
 quarkus.operator-sdk.generate-csv=true
 quarkus.container-image.builder=jib
 quarkus.operator-sdk.crd.validate=false
+
+# Operator config
+operator.operand.image=quay.io/keycloak/keycloak-x:latest
```
Comment (jonathanvila) [inline operator/src/main/resources/application.properties:7]: I would suggest another naming, as every element created by the operator are operands .... so probably operator.operand.keycloakImage could be clearer ? wdyt ?

File: operator/src/main/resources/example-postgres.yaml [hunk 23]
```diff
@@ -0,0 +1,39 @@
+# PostgreSQL StatefulSet
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: postgresql-db
```
Comment (jonathanvila) [inline operator/src/main/resources/example-postgres.yaml:5]: should we always prefix the operands with the cr name ? or at least with keycloak ? In the same namespace the user could have another postgresql-db for other purposes. Wdyt ?

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java (hunk 20 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:64]: Ok, unified the naming of the local variable with the field name.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java (hunk 21 above)
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/crds/KeycloakSpec.java]: 👍

File: operator/src/main/resources/application.properties (hunk 22 above)
Comment (vmuzikar) [reply operator/src/main/resources/application.properties]: Renamed to `operator.keycloak.image`. WDYT?

**Review by andreaTP — DISMISSED** [review]
LGTM

File: operator/src/main/resources/example-postgres.yaml (hunk 23 above)
Comment (jonathanvila) [reply operator/src/main/resources/example-postgres.yaml:5]: Yes you are right....I see what you mean.

**Review by jonathanvila — DISMISSED** [review]
LGTM :+1:

**Comment by vmuzikar** [comment]
I squashed the commits.

**Review by pedroigor — APPROVED** [review]

**Comment by AndrewFarley** [comment]
Please forgive me if this is a silly question, but can I ask if/when this will make it into a release of Keycloak-X?  I'd like to have a reasonable probe in Kubernetes instead of just a tcpProbe.  I can't see this being functional on the latest docker image, 16.1.1, or perhaps the question should instead be when will the docker images be updated eg: to the 17.0.1 release from yesterday?

**Comment by andreaTP** [comment]
Hi @AndrewFarley this operator will be included as preview in the next Keycloak release:
https://www.keycloak.org/2022/03/releases
